### PR TITLE
ICU-22452 API for resolving preferred IANA zone ID from a zone ID.

### PR DESCRIPTION
**Pull request by yumaoka**
##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22452
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/data/misc/timezoneTypes.txt
Comment (richgillam) [inline icu4c/source/data/misc/timezoneTypes.txt:48]: I see in the code what this new table is doing, but it wasn't clear to me from the rest of this PR what all the other changes to this file are doing or how they relate to the new API you're adding.  Maybe this is obvious to everybody who's spent a lot of time in the time-zone stuff, but would you mind explaining for the benefit of people like me who haven't?

**Review by richgillam — DISMISSED** [review]
You probably want to wait for review feedback from somebody who actually understands the time-zone stuff.  As for me, it all seems pretty straightforward, except for many of the changes in `timezoneTypes.txt` other than the new `ianaMap` resource.  I didn't know what all the other changes in that file were all about.

File: icu4c/source/data/misc/timezoneTypes.txt
Comment (richgillam) [reply icu4c/source/data/misc/timezoneTypes.txt:48]: Whoops.  Didn't see the "Incomplete" label until now.  Maybe that explains my questions...

Comment (yumaoka) [reply icu4c/source/data/misc/timezoneTypes.txt:48]: @richgillam 
>but it wasn't clear to me from the rest of this PR what all the other changes to this file are doing or how they relate to the new API you're adding.

To support the new data structure and produce a new table in ICU data files, we need to update cldr2icu tool. These changes found under tools were made for the data mapping. All other files in icu4c/icu4j are all related to 1) new data .txt 2) API and implementation 3) test cases.

Do you see something unrelated to this work?

Comment (richgillam) [reply icu4c/source/data/misc/timezoneTypes.txt:48]: My question was specifically about `timeZoneTypes.txt` and the changes to resources in it other than the new `ianaMap` resource.  Are the other changes in there somehow necessary to make the new `getIanaID()` functions work, or were they just other changes to the time-zone data in CLDR that came "along for the ride" when you re-ran the CLDR-to-ICU tools?  If they're part of the implementation of `getIanaID()`, I couldn't figure out what they were doing.

Comment (yumaoka) [reply icu4c/source/data/misc/timezoneTypes.txt:48]: OK. There are a few goals in CLDR-14453. First, sort out what set is appropriate for CLDR canonical zone IDs. And the second part is to provide mapping from CLDR canonical zone ID to IANA preferred zone ID. On CLDR side, we come up with new policy for a set of CLDR canonical zones. This was required to define IANA preferred zone ID mapping. This is the reason why you see a bunch of changes in timeZoneTypes.txt.

Comment (richgillam) [reply icu4c/source/data/misc/timezoneTypes.txt:48]: That makes sense.  Thank you for the explanation!  I guess those changes either don't result in a caller-visible change in behavior, or CLDR has judged any changes in behavior to be beneficial?

File: icu4c/source/i18n/ucal.cpp
Comment (FrankYFTang) [inline icu4c/source/i18n/ucal.cpp:635]: Why do we need to check  status == 0 here but not other methods? 
I only see ucal_getCanonicalTimeZoneID in this file perform such check.

File: icu4c/source/i18n/ucal.cpp
Comment (FrankYFTang) [inline icu4c/source/i18n/ucal.cpp:642]: I thnik we only need
```
    UnicodeString ianaID;
    return TimeZone::getIanaID(UnicodeString(id, len), ianaID, *status)
                  .extract(result, resultCapacity, *status);
```
Since both getIanaID and extract should check *status and return immediately anyway.

File: icu4c/source/i18n/timezone.cpp
Comment (FrankYFTang) [inline icu4c/source/i18n/timezone.cpp:1602]: ZoneMeta::getIanaID will set the returned ianaID to bogus, should we also set ianaID to bogus here for this error?

File: icu4c/source/i18n/unicode/timezone.h
Comment (FrankYFTang) [inline icu4c/source/i18n/unicode/timezone.h:466]: Should we also document that the return value could be bogus for error and "" if the input is "Etc/Unkown" ? or should both case return bogus string ?

File: icu4c/source/i18n/zonemeta.cpp
Comment (FrankYFTang) [inline icu4c/source/i18n/zonemeta.cpp:415]: since tmpIana is returned by ures_getStringByKey that is "a pointer to a zero-terminated UChar array which lives in a memory mapped/DLL file." Should we use the  ianaID.setTo(true, tmpIana, tmpLen);
to save some memory?
https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1UnicodeString.html#a9a2f115fb0975766d9b8823d98622295

File: icu4c/source/i18n/zonemeta.cpp
Comment (FrankYFTang) [inline icu4c/source/i18n/zonemeta.cpp:417]: Since the canonicalID is returned by the version which "overload method returns a persistent const char16_t*, which is guaranteed to persist    * (a pointer to a resource)."  of getCanonicalCLDRID() 

I think we should also use ianaID.setTo(true, canonicalID, -1);

https://github.com/unicode-org/icu/blob/cc2ddc0d113aa4db6efa6b071eca172e4a72f6cd/icu4c/source/i18n/zonemeta.h#L50

to save internal memory allocatioun for ianaID

File: icu4c/source/test/cintltst/ccaltst.c
Comment (FrankYFTang) [inline icu4c/source/test/cintltst/ccaltst.c:2849]: we should aslo test the case which the id is UNKNOWN and make sure the return of that is correct.

File: icu4c/source/test/intltest/tztest.cpp
Comment (FrankYFTang) [inline icu4c/source/test/intltest/tztest.cpp:2645]: as C test, we should also test the condition that id is UNKNOWN

File: icu4c/source/test/intltest/tztest.cpp
Comment (FrankYFTang) [inline icu4c/source/test/intltest/tztest.cpp:2653]: we need to test the return value of TimeZone::getIanaID and not assume the return value is always the same as ianaID from blackbox testing point of view. Maybe assert ianaID and the return value is always contain the same?

File: icu4c/source/test/intltest/tztest.cpp
Comment (FrankYFTang) [inline icu4c/source/test/intltest/tztest.cpp:2655]: we should also assert the return value is set to bogus here

File: icu4c/source/i18n/ucal.cpp
Comment (yumaoka) [reply icu4c/source/i18n/ucal.cpp:635]: There are other functions which also checks nullptr, such as _createTimeZone, ucal_getDefaultTimeZone, ucal_getHostTimeZone, and some others. Some of the does this differently. For example:

```
U_CAPI int32_t U_EXPORT2
ucal_getDefaultTimeZone(char16_t* result, int32_t resultCapacity, UErrorCode* ec) {
    int32_t len = 0;
    if (ec != nullptr && U_SUCCESS(*ec)) {
        TimeZone* zone = TimeZone::createDefault();
        if (zone == nullptr) {
            *ec = U_MEMORY_ALLOCATION_ERROR;
        } else {
            UnicodeString id;
            zone->getID(id);
            delete zone;
            len = id.extract(result, resultCapacity, *ec);
        }
    }
    return len;
}
```
I think ICU API should be consistent - maybe just not checking it and let the calling application crash. Anyway, we might discuss about our policy for this next week.

Comment (pedberg-icu) [reply icu4c/source/i18n/ucal.cpp:635]: Agree about check status == 0, per coding guidelines we should not be doing that: https://unicode-org.github.io/icu/userguide/dev/codingguidelines.html#sample-function-with-error-checking

File: icu4c/source/i18n/timezone.cpp
Comment (pedberg-icu) [reply icu4c/source/i18n/timezone.cpp:1602]: I think that is in the category of a "nice to have" consistency fix, but maybe it can be fixed later, should not block this PR.

Comment (yumaoka) [reply icu4c/source/i18n/timezone.cpp:1602]: Thanks for pointing out. I'll make this change.

File: icu4c/source/i18n/ucal.cpp
Comment (yumaoka) [reply icu4c/source/i18n/ucal.cpp:642]: Will check it again.

File: icu4c/source/i18n/unicode/timezone.h
Comment (yumaoka) [reply icu4c/source/i18n/unicode/timezone.h:466]: I will add a note in the API reference, explaining about "Etc/Unknown".

File: icu4c/source/i18n/zonemeta.cpp
Comment (yumaoka) [reply icu4c/source/i18n/zonemeta.cpp:415]: Thanks for the suggestion. Will do.

File: icu4c/source/test/cintltst/ccaltst.c
Comment (yumaoka) [reply icu4c/source/test/cintltst/ccaltst.c:2849]: Hmm.. it looks I missed the case. Will do.

File: icu4c/source/test/intltest/tztest.cpp
Comment (yumaoka) [reply icu4c/source/test/intltest/tztest.cpp:2645]: Yes.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/ZoneMeta.java
Comment (FrankYFTang) [inline icu4j/main/classes/core/src/com/ibm/icu/impl/ZoneMeta.java:72]: Why are these changes here? They are unrelated and unneeded for this new API, right?

File: icu4c/source/test/intltest/tztest.cpp
Comment (yumaoka) [reply icu4c/source/test/intltest/tztest.cpp:2655]: At this moment, it's not an API contact that bogus is set. Setting bogus is this implementation is for safety purpose. I'm OK add the check in the test case although.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/ZoneMeta.java
Comment (yumaoka) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/ZoneMeta.java:72]: By the standard Eclipse project setting. When you touch the file, it automatically optimize this to Java 7+ style.
We used to control coding style through Eclipse project file settings, but I know more people not using the env these days.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/ZoneMeta.java
Comment (FrankYFTang) [inline icu4j/main/classes/core/src/com/ibm/icu/impl/ZoneMeta.java:433]: how about we write write it in a way which has less condictional block in the function as:
```
if (canonicalID == null) return null;
try {
     UResourceBundle keyTypeData = UResourceBundle.getBundleInstance(ICUData.ICU_BASE_NAME,
                        "keyTypeData", ICUResourceBundle.ICU_DATA_CLASS_LOADER);
     UResourceBundle ianaMap = keyTypeData.get("ianaMap");
     UResourceBundle ianaTzMap = ianaMap.get("timezone");
     try {
         return ianaTzMap.getString(canonicalID.replace('/', ':'));
     } catch (MissingResourceException e) {
         // No IANA zone ID mapping. In this case, ianaId set by getCanonicalCLDRID()
         // is also a primary IANA id.
        return canonicalID;
     }
 } catch (MissingResourceException e) {
     // When CLDR canonical ID is not found for the input tzid.\
     return null;
 }
```

File: icu4c/source/i18n/ucal.cpp
Comment (pedberg-icu) [reply icu4c/source/i18n/ucal.cpp:642]: And if id == 0 || len == 0  then UnicodeString(id, len) will produce an empty string which will presumably cause TimeZone::getIanaID to fail with an error.

So yes, Frank's suggestion  probably simplifies things. But... the current implementation still works fine. Could be simplified in a future PR.

File: icu4c/source/i18n/unicode/timezone.h
Comment (pedberg-icu) [reply icu4c/source/i18n/unicode/timezone.h:466]: Since this has a UErrorCode parameter, not sure we want to make setting return to bogus part of the API contract.

File: icu4j/main/classes/core/src/com/ibm/icu/util/TimeZone.java
Comment (FrankYFTang) [inline icu4j/main/classes/core/src/com/ibm/icu/util/TimeZone.java:1190]: We have a 3 level nested if block here. Could we reduce that by 
```
        if (id == null || id.length() == 0 || id.equals(TimeZone.UNKNOWN_ZONE_ID)) {
            return ianaId;
        }
        String tmpIanaId = ZoneMeta.getIanaID(id);
        if (tmpIanaId == null) {
            return ianaId;
        }
        return tmpIanaId;
```

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/timezone/TimeZoneTest.java
Comment (FrankYFTang) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/timezone/TimeZoneTest.java:2409]: we should also check the case that the input is NKNOWN

File: icu4c/source/i18n/zonemeta.cpp
Comment (pedberg-icu) [reply icu4c/source/i18n/zonemeta.cpp:415]: Good idea

**Review by pedberg-icu — APPROVED** [review]
Frank has a lot of good suggestions for improvements in the ICU4C code. However, I think they can be done in  a follow-up PR, they would not change the basic operation of the code (except in some cases setting strings to bogus, which should not be part of the API contract for APIs that have a UErrorCode param. So I am approving this as is to unblock things...

**Comment by pedberg-icu** [comment]
> ....So I am approving this as is to unblock things...

Oh. But it seems the maven data did not get updated along with the ICU4J data jars, so the maven tests are failing. The maven data should be updated, see process in https://unicode-org.github.io/icu/processes/cldr-icu.html Process section 12

File: icu4c/source/data/misc/timezoneTypes.txt
Comment (yumaoka) [reply icu4c/source/data/misc/timezoneTypes.txt:48]: >I guess those changes either don't result in a caller-visible change in behavior, or CLDR has judged any changes in behavior to be beneficial?

Let me use an actual example. America/Thunder_Bay and America/Toronto. Previously, these zones are defined differently.
```
Zone America/Toronto	-5:17:32 -	LMT	1895
			-5:00	Canada	E%sT	1919
			-5:00	Toronto	E%sT	1942 Feb  9  2:00s
			-5:00	Canada	E%sT	1946
			-5:00	Toronto	E%sT	1974
			-5:00	Canada	E%sT
Zone America/Thunder_Bay -5:57:00 -	LMT	1895
			-6:00	-	CST	1910
			-5:00	-	EST	1942
			-5:00	Canada	E%sT	1970
			-5:00	Toronto	E%sT	1973
			-5:00	-	EST	1974
			-5:00	Canada	E%sT
```
Above definition is found in 2014g northamerica file. Both were in zone.tab, and some implementations show these two zones in the zone selection list. CLDR treated both of them as canonical zones. Since then, time zone database maintainer decided to make America/Thunder_Bay as "Link" of America/Toronto in backward file as below.
```
Link	America/Toronto		America/Thunder_Bay
```
The maintainer pointed out the difference before 1974 were not coming from reliable source, and concluded not worth keeping America/Thunder_Bay from America/Toronto. So he decided to move America/Thunder_Bay to backward file and use the rule same with America/Toronto.

ICU still support both America/Toronto and America/Thunder_Bay to create a time zone instance. Previously, getCanonicalID("America/Thunder_Bay") returns "America/Thunder_Bay, but with this change, it will be "America/Toronto".

In CLDR, there are two ways to define display names. One is through meta zone. Toronto and New_York share same zone display name like - "Eastern Time". In CLDR, we maintain the display name associated with meta zone America_Eastern, and maintain historic mapping between each canonical zone to meta zone. Another way to define display name is to assign names directly to canonical zone ID. The latter override the meta zone based display names.

So, deprecating America/Thunder_Bay as canonical zone and make it an alias of America/Toronto may change display names if there are zone based display name overrides (not by metazone and mapping), then display names will be changed. But assigning display names directly with zone ID is rare exception in CLDR at this moment. So practically, it should have no impacts.

Only possible change is - location based display name. In CLDR/ICU, date format patterm VVV shows exemplar city name. To produce a display name for VVV, CLDR maintains exemplar city translation. For example, "サンダーベイ" for America/Thunder_Bay in Japanese. To resolve the name, ICU gets a zone's canonical ID first, then look up the time zone exemplar city data. Previously it was America/Thunder_Bay, but it will be changed to America/Toronto. This is the actual impact.

From the time zone database maintainer's point of view, America/Thuder_Bay is only kept for backward compatibility purpose (therefore, it's defined in backward file). This means, time zone database consumer should not include America/Thunder_Bay in the zone selection list. For example, you can no longer select Thunder Bay as the system time zone in the current version of Ubuntu. VVV for America/Thunder_Bay produces "トロント" (Toronto) instead of "サンダーベイ" might not be great, but user is not supposed to select America/Thunder_Bay as system time zone any more.

File: icu4c/source/i18n/zonemeta.cpp
Comment (yumaoka) [reply icu4c/source/i18n/zonemeta.cpp:417]: will make this change.

File: icu4j/main/classes/core/src/com/ibm/icu/util/TimeZone.java
Comment (yumaoka) [reply icu4j/main/classes/core/src/com/ibm/icu/util/TimeZone.java:1190]: OK

**Comment by yumaoka** [comment]
> > ....So I am approving this as is to unblock things...
> 
> Oh. But it seems the maven data did not get updated along with the ICU4J data jars, so the maven tests are failing. The maven data should be updated, see process in https://unicode-org.github.io/icu/processes/cldr-icu.html Process section 12

I'm looking at maven one. I think it's related to data packaging, because ant build works.

File: icu4c/source/test/intltest/tztest.cpp
Comment (yumaoka) [reply icu4c/source/test/intltest/tztest.cpp:2654]: done

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/timezone/TimeZoneTest.java
Comment (yumaoka) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/timezone/TimeZoneTest.java:2410]: done

**Comment by yumaoka** [comment]
@pedberg-icu @FrankYFTang I think I fixed issues and updated code based on review feedback. Please review it again. Thanks.

File: icu4c/source/i18n/ucal.cpp
Comment (yumaoka) [reply icu4c/source/i18n/ucal.cpp]: @pedberg-icu Thanks for the link. I see a bunch of other code checking status == 0 or status != 0 along with U_FAILURE/U_SUCCESS. We may look at these code later.

File: icu4c/source/data/misc/timezoneTypes.txt
Comment (richgillam) [reply icu4c/source/data/misc/timezoneTypes.txt:48]: I see.  Thanks for the long explanation!  This makes sense, and I'm personally comfortable with what you're describing.

**Review by richgillam — APPROVED** [review]
LGTM.  Thanks for helping me understand what you're doing!

**Comment by pedberg-icu** [comment]
The copyright scan issue is fixed by https://github.com/unicode-org/icu/pull/2550, probably need to rebase to pick that up when it is merged. Working on that now.

**Comment by pedberg-icu** [comment]
> The copyright scan issue is fixed by #2550, probably need to rebase to pick that up when it is merged. Working on that now.

OK, I just merged #2550, so you can rebase this to pick up the missing copyright headers tha are blocking the tests for this.

**Comment by yumaoka** [comment]
> > The copyright scan issue is fixed by #2550, probably need to rebase to pick that up when it is merged. Working on that now.
> 
> @yumaoka OK, I just merged #2550, so you can rebase this to pick up the missing copyright headers that are blocking the tests for this.

Thanks @pedberg-icu. I rebased and squashed commits. Let's see if all CI checks passing.

**Review by pedberg-icu — APPROVED** [review]

**Comment by yumaoka** [comment]
@FrankYFTang I think I updated the code as you suggested. I think @pedberg-icu is waiting for this, so I will go ahead and merge. If something is still missing, then I can work on another PR. I'm going to merge this now.

File: icu4c/source/i18n/unicode/timezone.h
Comment (yumaoka) [reply icu4c/source/i18n/unicode/timezone.h:471]: done

File: icu4c/source/i18n/zonemeta.cpp
Comment (yumaoka) [reply icu4c/source/i18n/zonemeta.cpp]: done

File: icu4c/source/test/cintltst/ccaltst.c
Comment (yumaoka) [reply icu4c/source/test/cintltst/ccaltst.c:2850]: done

File: icu4c/source/test/intltest/tztest.cpp
Comment (yumaoka) [reply icu4c/source/test/intltest/tztest.cpp:2646]: done

File: icu4c/source/test/intltest/tztest.cpp
Comment (yumaoka) [reply icu4c/source/test/intltest/tztest.cpp:2656]: done. setToBogus, and test case also checks it.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/ZoneMeta.java
Comment (yumaoka) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/ZoneMeta.java]: done

File: icu4j/main/classes/core/src/com/ibm/icu/util/TimeZone.java
Comment (yumaoka) [reply icu4j/main/classes/core/src/com/ibm/icu/util/TimeZone.java]: done

File: icu4c/source/i18n/ucal.cpp
Comment (yumaoka) [reply icu4c/source/i18n/ucal.cpp]: Updated the C wrapper skip nullptr checking and directly calls TimeZone::getIanaID(). So sts==0 check is gone.

**Comment by pedberg-icu** [comment]
@yumaoka Many thanks for seeing this through!

**Pull request merged**
